### ISSUE TITLE
Training-Pipeline: Datensammlung, Labeling und Modell-Benchmarking

### ISSUE THREAD

**Issue by Nileneb** [Nileneb/MayringCoder#25]
## Ziel

Grundlage für späteres Fine-Tuning schaffen. Dazu muss zuerst **vergleichbare Rohdaten** über alle verfügbaren Modelle gesammelt und gelabelt werden — erst dann kann ein Training-Framework sinnvoll geplant werden.

## Warum vergleichbar statt --full

`--full` läuft so lange bis alles fertig ist — große Modelle brauchen dann 10x länger als kleine. Das ist kein fairer Vergleich.  
Stattdessen: **jedes Modell bekommt dasselbe Zeit-Budget** (~10 Minuten). Nach Ablauf wird graceful gestoppt. Dann vergleichen:  
- Wie viele Dateien hat jedes Modell in der Zeit geschafft?  
- Wie gut waren die Findings? (Parse-Erfolgsrate, Findings-Qualität)

---

### 1a. `--time-budget SECONDS` Flag in `checker.py`

Nach jeder analysierten Datei wird `time.perf_counter()` geprüft. Bei Überschreitung: graceful stop, Report trotzdem geschrieben.

```python
# In analyze_files() — nach jeder Datei:
if time_budget and (time.perf_counter() - start) > time_budget:
    print(f"  ⏱ Zeit-Budget ({time_budget}s) erreicht — stoppe nach {i+1} Dateien.")
    break
```

Ins Meta-JSON: `"time_budget_hit": true/false`.

### 1b. `benchmark.sh` — neues Script (trennen von `run-all.sh`)

`run-all.sh` ist für Produktion (vollständige Pipeline).  
`benchmark.sh` ist für Vergleich:

```bash
# Alle verfügbaren Ollama-Modelle automatisch einlesen:
MODELS=$(ollama list | tail -n +2 | awk '{print $1}')

# Jedes Modell mit gleichem Zeit-Budget:
for model in $MODELS; do
    python checker.py --model "$model" \
        --time-budget 600 \          # 10 Minuten
        --run-id "bench_$(slugify $model)" \
        --cache-by-model \
        --use-overview-cache \
        --use-turbulence-cache
done
```

Stufe 1 (Overview) und Stufe 2 (Turbulenz) laufen einmalig vorab — die sind unabhängig vom Analyse-Modell.

### 1c. `benchmark_summary.py` — Auswertungs-Script

Liest alle `bench_*`-Runs aus der SQLite-History und gibt eine Vergleichsmatrix aus:

```
Modell                    | Dateien | Findings | Parse-Fehler | Laufzeit | Score
--------------------------|---------|----------|--------------|----------|------
qwen2.5-coder:7b          |      18 |       42 |            1 |    587s  |  ★★★★
mistral:7b-instruct       |      12 |       28 |            3 |    603s  |  ★★★☆
deepseek-coder:6.7b       |      15 |       35 |            0 |    598s  |  ★★★★
llama3.2                  |       9 |       19 |            6 |    601s  |  ★★☆☆
```

Score-Formel (vorläufig): `(findings / files_analyzed) * (1 - parse_error_rate) * (high_confidence_ratio)`

---

### 2a. LLM-Call-Logger

In `_ollama_generate()` optional alle Calls in ein JSONL loggen:

```python
# Flag: --log-training-data
# Output: cache/<slug>_training_log.jsonl
{
  "timestamp": "...",
  "model": "qwen2.5-coder:7b",
  "filename": "app/Services/UserService.php",
  "prompt": "...",
  "raw_response": "...",
  "parsed_ok": true,
  "findings_count": 3,
  "run_id": "bench_qwen2-5-coder-7b"
}
```

Das ist der Rohstoff für Fine-Tuning: `prompt → response` Paare.

### 2b. Nur erfolgreiche Calls als Training-Kandidaten

Einfache Filterregel: `parsed_ok: true AND findings_count > 0` → Kandidat für Training.  
Schlechte Calls (`parsed_ok: false`) → Kandidat für "was soll das Modell NICHT machen".

---

### 3a. Automatisches Labeling (Heuristik)

Erster Pass ohne manuellen Aufwand:
- **Positiv**: Second-Opinion sagt BESTÄTIGT + Confidence high/medium
- **Negativ**: Second-Opinion sagt ABGELEHNT
- **Unklar**: kein Second-Opinion oder Confidence low → manuell reviewen

### 3b. Manuelles Labeling-CLI (`label.py`)

```
Finding #14 — app/Services/CreditService.php
Type: zombie_code | Severity: warning | Confidence: medium

Evidence:
> function calculateLegacyDiscount($amount) {

[B]estätigen  [A]blehnen  [P]räzisieren  [S]kip  [Q]uit
> _
```

Ergebnis wird direkt ins JSONL geschrieben:
```json
{"label": "positive", "human_reviewed": true, ...}
```

### 3c. Export für Fine-Tuning

```python
# export_training_data.py
# Input:  cache/<slug>_training_log.jsonl (gefiltert: label=positive)
# Output: training_data/finetune_<date>.jsonl

{"prompt": "<system>...</system>\n<user>Analysiere: ...</user>", "response": "{...JSON-Findings...}"}
```

Format kompatibel mit Unsloth / Axolotl / llama.cpp fine-tuning.

---

## Abhängigkeiten & Reihenfolge

```
Phase 1a: --time-budget Flag          ← sofort umsetzbar
Phase 1b: benchmark.sh                ← nach 1a
Phase 1c: benchmark_summary.py        ← nach 1b (braucht echte Daten)
Phase 2a: LLM-Call-Logger             ← parallel zu 1b
Phase 2b: Filter-Logik                ← nach 2a
Phase 3a: Auto-Labeling               ← nach 2b (Second-Opinion vorhanden)
Phase 3b: label.py CLI                ← nach 3a
Phase 3c: Export-Script               ← nach 3b
Fine-Tuning Framework                 ← nach 3c (separates Issue)
```

## Akzeptanzkriterien Phase 1 (MVP)

- [ ] `--time-budget SECONDS` stoppt graceful nach Zeitüberschreitung
- [ ] `benchmark.sh` liest verfügbare Ollama-Modelle automatisch ein
- [ ] Jedes Modell läuft isoliert (eigener run-id, cache-by-model)
- [ ] `benchmark_summary.py` gibt Vergleichsmatrix aus
- [ ] Meta-JSON enthält `time_budget_hit: true/false`

## Nicht in diesem Issue

- Fine-Tuning selbst (separates Issue nach Phase 3)
- Modell-Upload / Modelfile-Verwaltung
- Evaluation gegen Ground-Truth-Benchmark (kommt nach erstem Fine-Tuning)
